enhancements/user-workload-monitoring.md

```diff
@@ -0,0 +1,270 @@
+
+There are edge cases where an automated upgrade is not possible in the face of user workload monitoring.
+If users declare i.e. queries in ServiceMonitors which break in newer Prometheus versions
+we most likely cannot perform an automated migration. Such migration without knowledge of user intention behind each query would be exceptionally difficult.
+
+Instead, the "upgradable" condition in the ClusterOperator will be set to false.
+Routing configurations for alerts to be sent to specific teams need to be communicated/configured by an infrastructure owner using the existing Alertmanager stack that is part of cluster-monitoring.
+We will look into extension mechanisms post GA.
+With any OpenShift deployment, we only install the cluster-level part
+and allow infrastructure owner to "enable" the addition of user workload monitoring.
+
+### User Stories
+
+
+#### US1
+
+As a service owner, I’d like to configure a ServiceMonitor CRD so that Red Hat’s supported monitoring solution is able to scrape metrics for my service.
+
+#### US2
+
+As a service owner, I’d like to configure alerting rules for my service so that Red Hat’s supported monitoring solution notifies me if problems turn up.
+
+#### US3
+
+As a member of the operations team, I want to configure Prometheus one time so that all existing and any new namespaces will have access to user workload monitoring.
+
+#### US4
+
+As a member of the operations team, I'd like to add my own exporters so that I can monitor infrastructure services that are not monitored out-of-the-box.
+
+Note: This covers adding ServiceMonitor and PrometheusRule CRDs.
+
+#### US5
+
+As a developer working on a Red Hat product not part of the infrastructure, I’d like to integrate my metrics into a single, supported Prometheus stack so that they are available for my customers.
+
+#### US6
+
+As a service owner, I’d like to query the metrics I have access to from the OpenShift console.
+
+#### US7
+
+As a member of the operations tenant, I’d like to make sure that one tenant doesn't have access to metrics from another tenant.
+
+#### US8
+
+As a member of the operations team, I'd like the ability for application developers to access container cpu and memory usage metrics.
+
+### Implementation Details/Notes/Constraints
+
+#### Isolation
+
+The cluster monitoring as well as the user workload monitoring stack are composed of custom resources
+reconciled by two separate prometheus operators.
+Both prometheus operators are under supervision of cluster monitoring operator.
+This allows a separation of concerns in terms of upgradability, security settings, and scalability.
+The cluster monitoring Prometheus Operator continues to be deployed in the `openshift-monitoring` namespace.
+The user workload monitoring Prometheus Operator will be deployed in the `openshift-user-monitoring` namespace.
```

Comment:
By default?  Currently prom is 66% of total cluster resources in default install - is this going to double the size of a default cluster?  What’s the impact to footprint?

Reply:
user workload monitoring will not be enabled by default, so no..but i still think it's worth asking what the resource impact is when it's enabled.

Is prom alone 66%?  or the entire monitoring stack?  This proposal adds a second prom, but not a second alertmanager, for example.....

Reply:
Indeed, user workload monitoring is opt-in only and mutually exclusive with the OLM based prometheus operator and is not enabled by default. I will add this information to make it more clear.

Footprint: If the user opts into user workload monitoring, a second set of Prometheus servers will be installed. I will measure the idle footprint of those and include the measurements in this enhancements.

Reply:
is there any recommendation topology change for placement of the two prometheus setups?  if a user had configured infra nodes, should they configure two sets?

Reply:
@derekwaynecarr We will expose the same Prometheus configuration options as we do today for the cluster prometheus set. For placing those instances we currently expose node selector so users can place these on dedicated infra nodes. For completeness, here are the config options available today: https://github.com/openshift/cluster-monitoring-operator/blob/4e03b012c5e27c4080ac2d70f617d06f7746a1d8/pkg/manifests/config.go#L54-L63

Reply:
Would we turn this on by default in CI?  Footprint matters even if it’s not default for customers.

Reply:
i don't see why we would unless there are teams that think they could somehow benefit from using it.

i'd expect the monitoring team to have a job/tests that enable it and exercise it, of course.

Reply:
Agreed both to @smarterclayton and @bparees. We need good information on footprint and will for sure implement e2e tests in cluster monitoring operator. I also agree that this is not necessarily turned on by default in origin CI.